src/utils/url.js

```diff
@@ -44,10 +44,33 @@ export const getResource = (
 ): {| uri: string, headers?: { [string]: string } |} =>
   isUrlOnRealm(uri, auth.realm) ? getResourceWithAuth(uri, auth) : getResourceNoAuth(uri);
 
-export const hasProtocol = (url: string = '') => url.search(/\b(http|https):\/\//) !== -1;
+const protocolRegex = /^\s*((?:http|https):\/\/)(.*)$/;
+
+/** DEPRECATED */
+const hasProtocol = (url: string = '') => url.search(protocolRegex) !== -1;
+
+// Split a (possible) URL into protocol and non-protocol parts.
+// The former will be null if no recognized protocol is a component
+// of the string.
+//
+// Ignores initial spaces.
+/** PRIVATE -- exported only for testing */
+export const parseProtocol = (value: string): [string | null, string] => {
+  const match = protocolRegex.exec(value);
+  return match ? [match[1], match[2]] : [null, value];
+};
+
+/** DEPRECATED */
+export const fixRealmUrl = (url: string = '') => {
+  if (url === '') {
+    return '';
+  }
+  const trimmedUrl = url
+    .replace(/\s/g, '') // strip any spaces, internal or otherwise
+    .replace(/\/+$/g, ''); // eliminate trailing slashes
```

Comment:
This looks like a behavior change! Should be its own (tiny) commit separate from the refactors.

Certainly it's a helpful change -- see #3567 and in particular https://github.com/zulip/zulip-mobile/issues/3567#issuecomment-514844325 .

Reply:
It is, but after reconsideration, I'm not sure it's the right thing to do, at least on non-Zulip domains: I'm pretty sure it's possible to configure a(n admittedly bizarre) proxy server such that the extra slashes are needed.

I've left it out for now; it's an easy change.